Airbnb.Infrastructure/Service/DataMigrationService.cs

```diff
@@ -0,0 +1,155 @@
+        {
+            batch.Add(userExternal);
+
+            if (batch.Count >= batchSize)
+            {
+                await ProcessUserBatchAsync(batch, cancellationToken);
+                batch.Clear();
+            }
+        }
+
+        if (batch.Count > 0)
+            await ProcessUserBatchAsync(batch, cancellationToken);
+    }
+    
+    private async Task MigrateApartmentsAsync(string filePath, CancellationToken cancellationToken)
+    {
+        const int batchSize = 500;
+        var batch = new List<ExternalApartment>(batchSize);
+
+        await foreach (var apartmentExternal in _jsonReader.ReadApartment(filePath, cancellationToken))
+        {
+            batch.Add(apartmentExternal);
+
+            if (batch.Count >= batchSize)
+            {
+                await ProcessApartmentBatchAsync(batch, cancellationToken);
+                batch.Clear();
+            }
+        }
+
+        if (batch.Count > 0)
+            await ProcessApartmentBatchAsync(batch, cancellationToken);
+    }
+
+    private async Task ProcessApartmentBatchAsync(List<ExternalApartment> batch, CancellationToken cancellationToken)
+    {
+        var externalIds = batch.Select(a => a.ExternalId).ToList();
+
+        var existingIds = await _context.Apartments
+            .Where(a => externalIds.Contains(a.ExternalId))
+            .Select(a => a.ExternalId)
+            .ToListAsync(cancellationToken);
+
+        var newApartments = batch
+            .Where(a => !existingIds.Contains(a.ExternalId))
+            .Select(a => a.Adapt<Apartment>())
+            .ToList();
+
+        if (newApartments.Count == 0)
+        {
+            _logger.LogInformation("Batch skipped, all apartments already exist");
+            return;
+        }
+
+        _context.Apartments.AddRange(newApartments);
+        await _context.SaveChangesAsync(cancellationToken);
+    }
+    
+    private async Task ProcessUserBatchAsync(List<ExternalUser> batch, CancellationToken cancellationToken)
+    {
+        if (batch.Count == 0)
+            return;
+
+        var externalIds = batch.Select(u => u.ExternalId).ToList();
+        
+        var existingUserIds = await _userManager.Users
+            .Where(u => externalIds.Contains(u.ExternalId))
+            .Select(u => u.ExternalId)
+            .ToListAsync(cancellationToken);
+
+        foreach (var userExternal in batch)
+        {
+            if (existingUserIds.Contains(userExternal.ExternalId))
+                continue;
+
+            var newUser = userExternal.Adapt<ApplicationUser>();
+
+            var result = await _userManager.CreateAsync(newUser, DefaultPassword);
+
+            if (!result.Succeeded)
+            {
+                var errors = string.Join(", ", result.Errors.Select(e => e.Description));
+                _logger.LogError($"Error creating user {newUser.Email}: {errors}");
+            }
+        }
+    }
```

Comment:
You cannot rely on this logic. Imagine:
1. You got the file for migration from 28.05.25
2. You did migration
3. You got a new file at 28.06.25 with Updated info about some apartaments
4. Since you consider them migrated, you will skip them

My suggestion is implement upsert operation here and override in case they have the same external id